web/src/main/java/com/bazaarvoice/emodb/web/resources/databus/DatabusResource1.java

```diff
@@ -394,4 +413,89 @@ private SubjectDatabus getClient(BooleanParam partitioned) {
+                                        @QueryParam ("since") InstantParam sinceParam,
+                                        @Authenticated Subject subject) {
+        checkArgument(!Strings.isNullOrEmpty(table), "table is required");
+        Instant since = (sinceParam == null) ? null : sinceParam.get();
+        Date date = since != null ? Date.from(since) : null;
+        Iterator<UpdateRef> result = Iterators.emptyIterator();
+        int batchSize = 10_000;
+        Collection<String> splits = _dataStore.getSplits(table, batchSize);
+        splits.parallelStream().forEach(split -> {
+            Iterable<Map<String, Object>> documents = DataStoreStreaming.getSplit(_dataStore, table, split, true, consistency.get());
+            documents.forEach(document -> {
+                if (date.compareTo(Intrinsic.getLastUpdateAt(document)) <= 0) {
+                    List<UpdateRef> updateRefs = document.keySet().stream()
+                            .map(k -> getUpdateRef(table, k, document))
+                            .collect(Collectors.toList());
+                    _eventWriterRegistry.getDatabusWriter().writeEvents(updateRefs);
+                    Iterators.concat(result, updateRefs.iterator());
+                }
+            });
+        });
+        return streamingIterator(result);
+    }
+
+    @POST
+    @Path ("_replay-key-events/{table}/{key}")
+    //TODO
+//    @RequiresPermissions ("sor|read|{table}, system|raw_databus")
+    @Timed (name = "bv.emodb.sor.DatabusResource1.replay", absolute = true)
+    @ApiOperation (value = "Replay operation.",
+            notes = "Replay operation.",
+            response = Iterator.class
+    )
+    public Iterator<UpdateRef> replayEvents(@PathParam ("table") String table,
+                                        @QueryParam ("key") String key,
+                                        @QueryParam ("consistency") @DefaultValue ("STRONG") ReadConsistencyParam consistency,
+                                        @QueryParam ("since") InstantParam sinceParam,
+                                        @Authenticated Subject subject) {
+            checkArgument(!Strings.isNullOrEmpty(table), "table is required");
+            checkArgument(!Strings.isNullOrEmpty(key), "key is required");
+
+            Instant since = (sinceParam == null) ? null : sinceParam.get();
+            Date date = since != null ? Date.from(since) : null;
+            Map<String, Object> document = _dataStore.get(table, key, consistency.get());
+            Iterator<UpdateRef> result = null;
+            if (date.compareTo(Intrinsic.getLastUpdateAt(document)) <= 0) {
+                UpdateRef updateRef = getUpdateRef(table, key, document);
+                Set<UpdateRef> updateRefs = Collections.singleton(updateRef);
+                _eventWriterRegistry.getDatabusWriter().writeEvents(updateRefs);
```

Comment:
I would prefer moving this writing functionality to the databus interface and implementation in the DefaultDatabus. This way you don't have to inject the Writer Registry here, and avoid touching the EmoService.

Reply:
in that case we need to inject `DataStore dataStore` which is also bad, databaus shouldn't be aware about datastore.
databus has already had `writeEvents(updateRefs)` method that allows us to send events, so we just need to fetch `updateRefs` and pass.

Reply:
DefaultDatastore also have access to writeEvents(); check if that can help to move the functionality.

Reply:
yeah `DefaultDatastore` has access to `writeEvents()`, but our new method wont update records in datastore, it would just send events, so adding it to public interface will confuse in the future